src/coreclr/vm/codeman.h

```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+// This leaves 17 bits of the address to be handled by the RangeSectionFragment linked list
+//
+// For 32bit VA processes, use 1KB chunks holding pointers to the next level. This provides 8 bits of address resolution per level.    [31:24] and [23:16].
+// For the 32bit processes, only the last 16bits are handled by the RangeSectionFragment linked list.
+
+// Each level of the tree may be considered collectible or non-collectible (except for the top level, which is ALWAYS non-collectible)
+// The locking model of the tree structure is as follows.
+// Adding a newly allocated level requires holding the Reader lock. Multiple threads may add a level in parallel.
+// Removing a level requires holding the Writer lock.
+// No level which refers to a non-collectible fragment may be considered collectible.
+// A level may be upgraded from collectible to non-collectible, by changing the pointer which points at it to not have the sentinel bit.
+// A level may NOT ever be downgraded from non-collectible to collectible.
+// When a level becomes empty, it may be freed, and the pointer which points at it may be nulled out.
+//
+// Within the linked list of RangeSectionFragments, there are effectively 2 lists.
+// - The non-collectible list, which are always found first.
+// - The collectible list, which follows the non-collectible list.
+//
+// Insertion into the map uses atomic updates and fully pre-initialized RangeSection structures, so that insertions can be lock-free with regards to each other.
+// However, they are not lock-free with regards to removals, so the insertions use a Reader lock.
+//
+// Reading from the non-collectible data (the non-collectible portion of tree structure + the non-collectible list) does not require any locking at all.
+// Reading from the collectible data will require a ReaderLock. There is a scheme using the RangeSectionLockState where when there is an attempt to read without the
+// lock and we find collectible data, which will cause the runtime to upgrade to using the Reader lock in that situation.
+//
+// Reading this code you will ALSO find that the ReaderLock logic used here is intertwined with the GC mode of the process. In particular,
+// in cooperative mode and during GC stackwalking, the ReaderLock is always considered to be held.
+class RangeSectionMap
+{
+    class RangeSectionFragment;
+    class RangeSectionFragmentPointer;
+    typedef DPTR(RangeSectionFragment) PTR_RangeSectionFragment;
+    typedef DPTR(RangeSectionFragmentPointer) PTR_RangeSectionFragmentPointer;
+
+    // Helper structure which forces all access to the various pointers to be handled via volatile/interlocked operations
+    // The copy/move constructors are all deleted to forbid accidental reads into temporaries, etc.
+    class RangeSectionFragmentPointer
+    {
+    private:
+        TADDR _ptr;
+
+        TADDR FragmentToPtr(RangeSectionFragment* fragment)
+        {
+            TADDR ptr = (TADDR)fragment;
```

Comment:
```suggestion
            TADDR ptr = dac_cast<TADDR>(fragment);
```
or ifdef out the method for DACCESS_COMPILE